lib/matching/nearest_neighbour.cpp

```diff
@@ -18,29 +18,31 @@ std::vector<long> NearestNeighbour::operator()(
     const std::vector<ClusterChain> &chains) {
   // Prepare an empty vector that will store the index of each chain
   std::vector<long> matchIndex;
-  for (size_t i = 0; i < chains.size(); i++) {
-    // Initialize a maximum distance
+  for (size_t i = 0; i < descriptors.size(); i++) {
+    // Initialize a maximum distance and match index
     double nearest = std::numeric_limits<double>::max();
-    // In this case we are only really concerned with the last cluster.
-    // Get the frame number and the descriptor
-    auto descriptor = chains[i].descriptors().rbegin();
-    // Get the descriptor for that frame.
-    auto lastDescriptor = descriptor->second;
-    for (size_t j = 0; j < descriptors.size(); j++) {
-      // Compare the descriptor to the cluster. If it's closer than the previous
-      // chain then choose that one.
-      // TODO This means that there may be multipel chains pointing to the
-      // same descriptor.
-      long match = 0;
-      double distance = descriptors[j]->compare(lastDescriptor.get());
-      if (nearest > distance) {
-        // Better match than previous chain. Pick that one.
-        match = i;
-        nearest = distance;
+    long match = -1;
+    for (size_t j = 0; j < chains.size(); j++) {
+      // Check if we already assigned this chain
+      if (std::find(matchIndex.begin(), matchIndex.end(), j) ==
```

Comment:
Probably takes O(n). The usage of an additional map `std::map<chainIndex, descriptorIndex>` (or std::unordered_map) could reduce it to O(log n) add the added cost of additional memory usage and a slightly more complicated `(nearest > distance)` update. But ok for the moment.